### ISSUE TITLE
TRAINING HARNESS + SELF-PLAY PIPELINE

### ISSUE THREAD

**Issue by athariandre** [athariandre/yasta#3]
“Implement a reproducible, deterministic training pipeline with self-play, match simulation, reward shaping, and model checkpointing.”
✅ OVERVIEW

This PR adds a new directory called training/ which holds the entire reinforcement-learning training stack.
Nothing in agent.py changes yet — this PR focuses entirely on:

A local simulation loop for self-play

A reward function

A match runner abstraction

A PPO-compatible rollout generator

Deterministic environment reset and stepping

Data logging + checkpointing

This ensures you can run training independently of the Flask server and produce model weights that future PRs (PR-3/PR-4) will teach the live agent to load.

✅ GOALS

The code generated by this PR must meet these goals:

Deterministic, reproducible local training

Self-play between two copies of the current heuristic agent

A clean environment API

A rollout collector abstraction

Reward function aligned with Tron survival logic

Model checkpoints saved to /training/checkpoints/

No dependency beyond CPU-safe PyTorch

No extra Docker size

Configurable hyperparameters

Runs fully offline (no Flask, no HTTP)

✅ MUST CREATE THESE FILES

Inside a new folder:

training/
    env.py
    game_sim.py
    rewards.py
    rollout.py
    trainer.py
    config.py
    readme.md
    checkpoints/   (empty, gitkeep)

Do NOT modify the competition files except when directly reading game classes.

Each file’s responsibilities are defined below.

✅ FILE-BY-FILE SPECIFICATION
✅ training/config.py

A simple configuration file containing:

BOARD_H, BOARD_W

PPO hyperparameters

rollout length

batch size

learning rate

discount factor

reward weights

seed

checkpoint interval

max training steps

Everything should be inside a dictionary-like class or namedtuple.

Nothing fancy.

✅ training/game_sim.py

This file provides a local simulation wrapper over the official case_closed_game.py classes.

Must implement:
class LocalGame:
    def __init__(self, seed: int): ...
    def reset(self): ...
    def step(self, action_p1, action_p2): ...
    def get_obs(self, player_number): ...
    def is_done(self): ...
    def clone(self): ...  # optional but preferred

Requirements:

Uses the official Game, Agent, Direction classes

Wraps game.step() exactly, without altering official logic

Does not depend on Flask

Exposes a tuple: (obs, reward, done, info)

Observations must include:

board grid

both head positions

trails

boosts

turn count

previous direction

Observations must be numpy arrays (CPU only)

✅ training/rewards.py

Reward shaping module.

Must implement these terms explicitly:

Survival Reward

+1 for each turn alive

+100 for win

−100 for loss

0 for draw

Space Control Reward

+0.01 × (Voronoi margin / total_cells)

Mobility Reward

+0.002 × reachable_area

Trap Avoidance Penalty

−0.05 if degree(final_head) ≤ 1

Head-on avoidance

−1 if predicted collision

Boost Efficiency

+0.1 when boost increases reachable_area

−0.1 when boost decreases it

Provide a function:
def compute_reward(prev_state, action, next_state, done, outcome):
    ...

Reward must be scalar float.

✅ training/env.py

Implements a unified RL-style environment:

class TronEnv:
    def __init__(self, config): ...
    def reset(self): ...
    def step(self, action_self): ...

Requirements:

This is a single-agent wrapper using self-play

Opponent = copy of agent OR heuristic agent

Must support:

self-vs-self training

self-vs-heuristic training

Maintains symmetric self-play:

50% of games as player 1

50% as player 2

Converts action indices → directions ("UP", etc)

✅ training/rollout.py

Class for generating PPO rollouts:

class RolloutCollector:
    def __init__(...):
    def collect(env, policy):
         # returns batches of:
         # obs, actions, logprobs, rewards, dones, values

Requirements:

Must store sequences aligned for PPO

Must correctly handle episode termination

Must preallocate memory for speed

No GPU calls

Everything CPU tensors

✅ training/trainer.py

A training script that:

Loads config

Initializes:

Policy model

Rollout collector

Environment

Trains PPO for X updates

Saves checkpoints every N updates

Logs:

win rate

average game length

average reward

loss terms

Seeds all RNG paths

Model:

PR-2 must NOT implement a neural network yet.
Instead, create a placeholder:

class DummyPolicy:
    def act(self, obs): 
         # returns:
         #   action index (0–3)
         #   logprob (fake)
         #   value estimate (fake)

The real policy NN comes in PR-3.

But all training code must already be compatible with:

policy.act()

policy.evaluate_actions()

So when PR-3 swaps in a real model, no changes are needed elsewhere.

✅ training/readme.md

Document:

How to run training

How to evaluate checkpoints

How to load a checkpoint

What each module does

How to integrate with agent.py in future PRs

✅ NO CHANGES IN THIS PR:

✅ Do NOT modify agent.py
✅ Do NOT modify the judge
✅ Do NOT modify the engine logic
✅ Do NOT add heavy ML libraries
✅ Only CPU PyTorch allowed

✅ ACCEPTANCE CRITERIA

PR-2 passes review if:

✅ All files listed above exist
✅ No file modifies competition code or agent.py
✅ Rollout collector structured correctly
✅ Environment produces consistent self-play episodes
✅ Rewards are implemented exactly as specified
✅ Placeholder policy works and training script runs end-to-end
✅ Only CPU PyTorch is used
✅ Entire training runs without Flask
✅ No model inference in agent.py yet
